README
NAME
    Xango - Event Based High Performance Web Crawler Framework

SYNOPSIS
      use Xango;

DESCRIPTION
      ******************  WARNING! WARNING! WARNING! ****************
      * This new release is NOT compatible with the old Xango 0.0x! *
      * Beware and use at your own risk.                            *
      ***************************************************************

    Xango is a framework for writing web crawlers. As such, it doesn't do a
    whole lot by itself - you need to create custom handlers to do the grunt
    work. See the documentation for Xango::Manual::Intro for more details on
    how to write your own crawler.

    Please note that Xango is still in beta. Some behavior may change as we
    keep on developing it.

COMPONENTS
    The main component that comes with Xango is the Broker component. The
    Broker handles the basic flow of a web crawler - it accepts data to
    fetch, applies some policies to it, fetches it, and then sends it for
    final processing.

    The only concrete implementation provided by the Broker of the above
    flow is the part where the requested URI is fetched. You must provide
    the rest of the logic. See Xango::Broker for details

    The Handler is the component in which the Broker delegates the above
    processing to.

CONSTANTS
    There are a few constants that may change the behavior of Xango. You can
    either set these by pre-declaring the constant subroutine of the same
    name or you can set them by populating the corresponding environment
    variable, which is prefixed with 'XANGO_'

    For example,

      BEGIN { sub Xango::DEBUG { 1 } }
      use Xango;

    or

      $ENV{XANGO_DEBUG} = 1;
      use Xango;

    will both toggle the DEBUG switch.

    See blow for the list of possible flags and switches.

  DEBUG
    Set this to 1 if you want to enable debugging messages.

FUNCTIONS
  debug
  info
    These are simple interface to the underlying logging mechanism

DEBUGGING
    To turn debugging on, you need to pre-declare some constants (which, in
    Perl, are subroutines). For example, to turn debugging on, you need to
    say something like this:

      sub Xango::DEBUG { 1 }
      use Xango;

    Xango will recognize that the DEBUG flag is already set, and turn
    debugging on.

BUGS
    Plenty, I'm sure. Please report bugs to RT
    http://rt.cpan.org/NoAuth/Bugs.html?Dist=Xango

TODO
    Documentation.
        Documentation for this distribution is half-baked at best. It needs
        a lot of work, including a tutorial.

CAVEATS
    This is an alpha version release. Expect a lot of things to be still
    either in broken or experimental stage.

SEE ALSO
    POE Xango::Manual::Intro Xango::Broker::Pull Xango::Broker::Push

AUTHOR
    Copyright 2005-2006 Daisuke Maki <dmaki@cpan.org>. All rights reserved.
    Development funded by Brazil, Ltd. <http://b.razil.jp>



Changes
Changes
=======

Except where noted, all changes made by Daisuke Maki <dmaki@cpan.org>

1.08_01 - 16 May 2006
 - Add clear() call to the DNS cache on startup. On systems with prolonged
   process life time, it's likely that the DNS cache is holding bad data,
   so we might as well reinitialize.
 - add tools/checkreqs.pl and custom Makefile.PL
 - Fix Push.pm so that shutdown_broker state kills alarms.

 [EXPERIMETAL]
 - Try to work with servers that send 400/408 upon too many connections
   (the sysadmins must hate me) by gracefully reducing the number of
   HTTP fetchers upon encountering those return status

1.08 - 01 May 2006
 - Implemented the "Loop" parameter in the Push crawler. This will allow
   you to keep the broker session even without a pending job queue.

1.07 - 26 Apr 2006
 - Hiroyuki Nishimura makes me realize that there's no way to shutdown
   a Push broker. Also added a shutdown_broker state, so you can easily
   shutdown the broker.

1.06 - 17 Apr 2006
 - Add EnableDnsCache argument so that you can enable/disable DNS
   caching. This is on by default.
 - Make Cache::Cache a requirement. 

1.05 - 12 Apr 2006
 - Fix constructor to check for parameters passed to it (other than the
   config file). Now Xango::Broker::* constructors properly use
   config file as the defaults, and then let other parameters override 
   those values.

1.04 - 01 Apr 2006
 - Add missing SimplePull.pm for testing (Reported by Tatsuhiko Miyagawa)
 - Fix POD

1.03 - 01 Apr 2006
 - Fix the push crawler: it was spawning way too many http connections
   when a massive amount of URI is enqueue'd to the broker.
   (Reported by Yappo)

1.02 - 04 Feb 2006 
 - properly call 'prep_request' on the handler, not the broker session

1.01 - 26 Jan 2006 ("Bang My Head Against The Wall" Release)
 - Fix the next state that fake_error_response posts to.

1.00 - 12 Dec 2005
 - Xango::Job can accept simple scalars, not just URI objects
   (nudged by Tasuku Suenaga over some prime ribs)
 - Only run POD tests when disttest is specified
 - Cleanup

0.99_02 - 18 Oct 2005
 - Oops, Xango::Tutorial should have been Xango::Manual::Intro
 - Document Xango::Broker::Base constructor arguments
 - Fix initialization. First load from config file, then use arguments.
 - Silence Pod::Test::Coverage (but not quite done yet)

0.99_01 - 17 Oct 2005
 *** Backwards Incompatible Changes ***
 - Change structure drastically: Xango::Broker is now separated into
   Xango::Broker::Pull, Xango::Broker::Push.
 - States have changed -- beware!

 *** House Cleaning ***
 - 0.99 no longer resides on the same svn repository as 0.0x.

 - Handler component's alias no longer needs to be 'handler'. You may specify
   arbitrary names in the Xango::Broker constructor.
 - Broker components' alias no longer needs to be 'broker'. Same as above.
 - Add Xango::Manual::Intro
 - apply_policy is now called synchronously, not thought POE's FIFO
 - Allow toggling DEBUG switch from $ENV via $ENV{XANGO_DEBUG}
 - Add tests for push and pull crawlers

0.09 - 
 - Add a 'shutdown' state.

0.08 - 09 Jun 2005
 - Fix POE::Kernel->sig() usage.
 - Autoflush debug handle
 - Add DnsCacheArgsDeref config parameter, in case you need to pass
   non-reference parameter list to the cache constructor. I needed this
   to use Cache::FastMmap as the cache engine

0.07 - 22 May 2005
 - Doc tweaks.
 - Properly use DnsComponentClass
 - Properly use DNS_COMP_ALIAS
 - Fix Xango::Config::init()

0.06 - 20 May 2005
 - Build.PL tweaks

0.05 - 08 May 2005
 - Remove processors from distribution.
 - Doc tweaks.

0.04 - 24 Apr 2005

 - Add HttpComponentClass, HttpComponent, JobRetrievalDelay, ReloadConfig
   to config parameters
 - Add Log::Dispatch

0.03 - 24 Mar 2005
  - rename xago-broker to xango (no need, as Xango has been integrated
    into one process, rather than multiple)
  - remove unneeded prerequisites

0.02 - 17 Mar 2005
  - Configuration variables noted
  - handle sigpipe in case remote servers kill their connection
    (PoCo::Component::HTTP does not handle this)

0.01 - 07 Feb 2005
  - Renamed to Xango from Brazil-Crawler

lib/Xango.pm
# $Id: Xango.pm 107 2006-05-17 09:01:19Z daisuke $
#
# Copyright (c) 2005 Daisuke Maki <dmaki@cpan.org>
# All rights reserved.

package Xango;
use strict;
use vars qw($VERSION $LOGDISPATCH);
use IO::Handle;
use Log::Dispatch;
use Log::Dispatch::Handle;
use POSIX();

BEGIN
{
    $VERSION = '1.08';

    my $default_from_env = sub {
        my($var, $default) = @_;
        my $env_var = 'XANGO_' . $var;
        return exists $ENV{$env_var} ? $ENV{$env_var} : $default;
    };
    # Define symbols. If users define them before Xango.pm is loaded,
    # then we respect their values.
    my %constants = map { ($_ => $default_from_env->($_, 0)) } qw(DEBUG);

    while (my($const, $value) = each %constants) {
        if (! __PACKAGE__->can($const)) {
            eval "sub $const { '$value' }";
            die if $@;
        }
    }

    if (!$LOGDISPATCH) {
        $LOGDISPATCH = Log::Dispatch->new(callbacks => \&_format4logdispatch);
        if (&DEBUG) {
            my $io     = IO::Handle->new();
            my $dup    = $io->fdopen(fileno(STDERR), "w");
            $dup->autoflush(1);
            my $handle = Log::Dispatch::Handle->new(
                name      => 'xgDEBUG',
                min_level => 'debug',
                handle    => $dup,
            );
            $LOGDISPATCH->add($handle);
        }
    }
}

sub info
{
    if ($LOGDISPATCH) {
        $LOGDISPATCH->log(level => 'info', message => "@_");
    }
}

sub debug
{
    if(DEBUG && $LOGDISPATCH) {
        $LOGDISPATCH->log_to(level => 'debug', name => 'xgDEBUG', message => "@_");
    }
}

sub _format4logdispatch
{
    my %args    = @_;
    my $message = $args{message};
    $message =~ s/\n$//;
    return
        join('', POSIX::strftime("%Y%m%d%H%M%S ", localtime), $message, "\n");
}


1;

__END__

=head1 NAME

Xango - Event Based High Performance Web Crawler Framework

=head1 SYNOPSIS

  use Xango;

=head1 DESCRIPTION

  ******************  WARNING! WARNING! WARNING! ****************
  * This new release is NOT compatible with the old Xango 0.0x! *
  * Beware and use at your own risk.                            *
  ***************************************************************

Xango is a framework for writing web crawlers. As such, it doesn't do
a whole lot by itself - you need to create custom handlers to do the
grunt work. See the documentation for L<Xango::Manual::Intro|Xango::Manual::Intro>
for more details on how to write your own crawler.

Please note that Xango is still in beta. Some behavior may change as we
keep on developing it.

=head1 COMPONENTS

The main component that comes with Xango is the Broker component.
The Broker handles the basic flow of a web crawler - it accepts data
to fetch, applies some policies to it, fetches it, and then sends it 
for final processing. 

The only concrete implementation provided by the Broker of the above flow 
is the part where the requested URI is fetched. You must provide the rest
of the logic. See L<Xango::Broker|Xango::Broker> for details

The Handler is the component in which the Broker delegates the above
processing to.

=head1 CONSTANTS

There are a few constants that may change the behavior of Xango. You can
either set these by pre-declaring the constant subroutine of the same name
or you can set them by populating the corresponding environment variable,
which is prefixed with 'XANGO_'

For example,

  BEGIN { sub Xango::DEBUG { 1 } }
  use Xango;

or 

  $ENV{XANGO_DEBUG} = 1;
  use Xango;

will both toggle the DEBUG switch.

See blow for the list of possible flags and switches.

=head2 DEBUG

Set this to 1 if you want to enable debugging messages. 

=head1 FUNCTIONS

=head2 debug

=head2 info

These are simple interface to the underlying logging mechanism

=head1 DEBUGGING

To turn debugging on, you need to pre-declare some constants (which, in Perl,
are subroutines). For example, to turn debugging on, you need to say something
like this:

  sub Xango::DEBUG { 1 }
  use Xango;

Xango will recognize that the DEBUG flag is already set, and turn debugging on.

=head1 BUGS

Plenty, I'm sure. 
Please report bugs to RT http://rt.cpan.org/NoAuth/Bugs.html?Dist=Xango

=head1 TODO

=over 4

=item Documentation. 

Documentation for this distribution is half-baked at best. It needs a lot of
work, including a tutorial.

=back

=head1 CAVEATS

This is an alpha version release. Expect a lot of things to be still either
in broken or experimental stage.

=head1 SEE ALSO

L<POE|POE>
L<Xango::Manual::Intro>
L<Xango::Broker::Pull>
L<Xango::Broker::Push>

=head1 AUTHOR

Copyright 2005-2006 Daisuke Maki E<lt>dmaki@cpan.orgE<gt>.
All rights reserved.
Development funded by Brazil, Ltd. E<lt>http://b.razil.jpE<gt>

=cut